com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -104,29 +116,166 @@ public int AddEntry(ulong networkObjectId, int behaviourIndex, int variableIndex
             return pos;
         }
 
+        /// <summary>
+        /// Write an Entry to send
+        /// Must match ReadEntry
+        /// </summary>
+        /// <param name="writer">The writer to write the entry to</param>
+        internal void WriteEntry(NetworkWriter writer, in Entry entry)
+        {
+            //todo: major refactor.
+            // use blittable types and copy variable in memory locally
+            // only serialize when put on the wire for network transfer
+            writer.WriteUInt64(entry.Key.NetworkObjectId);
+            writer.WriteUInt16(entry.Key.BehaviourIndex);
+            writer.WriteUInt16(entry.Key.VariableIndex);
+            writer.WriteUInt16(entry.Key.TickWritten);
+            writer.WriteUInt16(entry.Position);
+            writer.WriteUInt16(entry.Length);
+        }
+
+        /// <summary>
+        /// Read a received Entry
+        /// Must match WriteEntry
+        /// </summary>
+        /// <param name="reader">The readed to read the entry from</param>
+        internal Entry ReadEntry(NetworkReader reader)
+        {
+            Entry entry;
+            entry.Key.NetworkObjectId = reader.ReadUInt64();
+            entry.Key.BehaviourIndex = reader.ReadUInt16();
+            entry.Key.VariableIndex = reader.ReadUInt16();
+            entry.Key.TickWritten = reader.ReadUInt16();
+            entry.Position = reader.ReadUInt16();
+            entry.Length = reader.ReadUInt16();
+            entry.Fresh = false;
+
+            return entry;
+        }
+
         /// <summary>
         /// Allocate memory from the buffer for the Entry and update it to point to the right location
         /// </summary>
         /// <param name="entry">The entry to allocate for</param>
         /// <param name="size">The need size in bytes</param>
-        public void AllocateEntry(ref Entry entry, long size)
+        public void AllocateEntry(ref Entry entry, int index, int size)
         {
             // todo --M1--
             // this will change once we start reusing the snapshot buffer memory
             // todo: deal with free space
             // todo: deal with full buffer
 
-            entry.Position = (ushort)FreeMemoryPosition;
+            int pos;
+
+            if (entry.Length > 0)
+            {
+                Allocator.Deallocate(index);
+            }
+
+            bool ret = Allocator.Allocate(index, size, out pos);
+
+            if (!ret)
+            {
+                //todo: error handling
```

Comment:
Maybe throw an assert in here for the time being?  This is a 'game over' situation, yes?

Reply:
Would you clarify what you would like to see? Something like:

```namespace MLAPI.Exceptions
{
    public class SnapshotMemoryException : Exception``` ?

Reply:
I read `throw` as in exception. The alternative is indeed just an assert